splash/browser_tab.py

```diff
@@ -768,3 +838,146 @@ def log(self, message, min_level=None):
+    Python callbacks.
+
+    It is "one shot" because either `resume()` or `error()` should be called
+    exactly _once_. It raises an exception if the combined number of calls
+    to these methods is greater than 1.
+
+    If timeout is zero, then the timeout is disabled.
+    """
+
+    def __init__(self, parent, callback, errback, timeout=0):
+        self.name = str(uuid.uuid1())
+        self._used_up = False
+        self._callback = callback
+        self._errback = errback
+
+        if timeout < 0:
+            raise ValueError('OneShotCallbackProxy timeout must be >= 0.')
+        elif timeout == 0:
+            self._timer = None
+        elif timeout > 0:
+            self._timer = QTimer()
+            self._timer.setSingleShot(True)
+            self._timer.timeout.connect(self._timed_out)
+            self._timer.start(timeout * 1000)
+
+        super(OneShotCallbackProxy, self).__init__(parent)
+
+    @pyqtSlot()
+    @pyqtSlot('bool')
+    @pyqtSlot('double')
+    @pyqtSlot('QByteArray')
+    @pyqtSlot('QVariantMap')
+    def resume(self, value=None):
+        if self._used_up:
+            raise OneShotCallbackError("resume() called on a one shot" \
+                                       " callback that was already used up.")
+
+        self._use_up()
+        self._callback(self._qt_object_to_py_object(value))
+
+    @pyqtSlot(str)
+    def error(self, message):
+        if self._used_up:
+            raise OneShotCallbackError("error() called on a one shot" \
+                                       " callback that was already used up.")
+        self._use_up()
+        self._errback(str(message))
+
+    def cancel(self, reason):
+        self._use_up()
+        self._errback("One shot callback canceled due to: %s." % reason)
+
+    def _dict_looks_like_list(self, dict_):
+        """
+        QT converts JS arrays and objects into QVariantMap, which it then
+        converts to Python dict.
+
+        The problem with this behavior is that we don't know if the dict
+        we received started life as a JS Array or a generic JS object. We
+        use a simple heuristic to determine if dict_ "looks like" a
+        list: the dict_ is a list if dict_.keys() contains all of the integers
+        in the range [0,n-1], where n is the number of keys in the dict_.
+        """
+
+        if len(dict_) == 0:
+            return False
+
+        try:
+            indices = map(int, dict_.keys())
+        except ValueError:
+            return False
+
+        if min(indices) != 0 or max(indices) != len(indices) - 1:
+            return False
+
+        return True
+
+    def _dict_to_list(self, dict_):
+        """
+        Convert a list-like dictionary to a list.
+
+        See _dict_looks_like_list().
+        """
+
+        list_ = list()
+
+        for i in range(0, len(dict_)):
+            list_.append(dict_[str(i)])
+
+        return list_
+
+    def _qt_object_to_py_object(self, qt_object):
```

Comment:
There is [splash.qtutils.qt2py](https://github.com/scrapinghub/splash/blob/de93746525e1f92e1bfe3112ed4915fc6dd54497/splash/qtutils.py#L158) function. Do you think it can be reused (or improved and reused)?